<?xml version="1.0"?>
<launch>
    <!-- ADI 3D ToF Sensor -->

    <!-- General arguments -->

    <!-- Prefix -->
    <arg name="ns_prefix_cam1" default="cam1" />
    <!-- Input mode
        0:Real Time Sensor
        1:Reserved
        2:Rosbag bin
        3:Host connected to ADTF31xx
    -->
    <arg name="arg_input_sensor_mode"  default="3"/>
    <!-- IP of the ADTF31xx board : Applicable only if the input mode is 3 -->
    <arg name="arg_input_sensor_ip"  default="10.43.0.1"/>
    <!-- Input filename : Applicable only if the input mode is 2 -->
    <arg name="arg_in_file_name" default="$(find adi_3dtof_adtf31xx)/../adi_3dtof_input_video_files/adi_3dtof_height_170mm_yaw_0degrees_cam3.bin" />
    <!-- Height of the sensor from camera in meters -->
    <arg name="arg_camera_height_from_ground_in_mtr" default="0.15" />
    <!-- abThreshold -->
    <arg name="arg_ab_threshold" value="10" />
    <!-- confidenceThreshold -->
    <arg name="arg_confidence_threshold" value="10" />
    <!-- Enable RVL compression for depth and ab images -->
    <arg name="arg_enable_depth_ab_compression" default="0"/>
    <!-- Enable Point cloud publishing -->
    <arg name="arg_enable_point_cloud_publish" default="0"/>
    <!-- Encoding Type
        "mono16" - 16 bit sincle color channel
        "16UC1" - 16 bit unsigned integer
     -->
    <arg name="arg_encoding_type" value="mono16" />
    <!-- Configuration file name of ToF SDK 
         config_adsd3500_adsd3100.json - For ADSd3100(MP sensor)
         config_adsd3500_adsd3030.json - For ADSd3030(VGA sensor)
    -->
    <arg name="arg_config_file_name_of_tof_sdk" value="config/config_adsd3500_adsd3100.json" />
    <!-- camera_mode
    MP(1024x01024) sensor:
        "sr-native" - 0
        "lr-native" - 1
        "sr-qnative" - 2
        "lr-qnative" - 3
        "sr-mixed" - 5
        "lr-mixed" - 6

    VGA(640x512) sensor:
        "sr-native" - 0
        "lr-native" - 1
        "lr-qnative" - 3
        "sr-mixed" - 5
        "lr-mixed" - 6
    -->
    <arg name="arg_camera_mode" value="3" />

    <!-- Arguments on TF-related info  -->
    <!-- Camera device frame name -->
    <arg name="cam1_base_frame_optical" default="$(eval arg('ns_prefix_cam1') + '_adtf31xx_optical')" />
    <!-- Camera optical frame name -->
    <arg name="cam1_base_frame" default="$(eval arg('ns_prefix_cam1') + '_adtf31xx')" />
    <!-- map
          ^
          |
      camera_device
          ^
          |
      camera_optical
    -->
    <arg name="cam1_parent_frame" value="map"/>
    <arg name="cam1_child_frame" value="$(arg cam1_base_frame_optical)"/>

    <!-- Optical to Device rotation, this are standard values, would never change -->
    <arg name="cam_optical_to_base_roll" value="-1.57"/>
    <arg name="cam_optical_to_base_pitch" value="0"/>
    <arg name="cam_optical_to_base_yaw" value="-1.57"/>

    <!-- Camera position wrt map -->
    <arg name="cam1_pos_x" value="0.0"/>
    <arg name="cam1_pos_y" value="0.0"/>
    <arg name="cam1_pos_z" value="$(arg arg_camera_height_from_ground_in_mtr)"/>
    <arg name="cam1_roll" value="0.0"/>
    <arg name="cam1_pitch" value="0.0"/>
    <arg name="cam1_yaw" value="0.0"/>

    <!-- Main Node-->
    <group ns="$(arg ns_prefix_cam1)">
        <node name="adi_3dtof_adtf31xx_node" pkg="adi_3dtof_adtf31xx" type="adi_3dtof_adtf31xx_node" cwd="node" output="screen" >
            <param name="param_camera_link" value="$(arg cam1_base_frame_optical)"/>
            <param name="param_input_sensor_mode" value="$(arg arg_input_sensor_mode)" />
            <param name="param_input_sensor_ip" value="$(arg arg_input_sensor_ip)" />
            <param name="param_input_file_name" value="$(arg arg_in_file_name)"/>
            <param name="param_config_file_name_of_tof_sdk" value="$(arg arg_config_file_name_of_tof_sdk)" />
            <param name="param_camera_mode" value="$(arg arg_camera_mode)"/>
            <param name="param_enable_depth_ab_compression" value="$(arg arg_enable_depth_ab_compression)"  />
            <param name="param_ab_threshold" value="$(arg arg_ab_threshold)"/>
            <param name="param_confidence_threshold" value="$(arg arg_confidence_threshold)" />
            <param name="param_enable_point_cloud_publish" value="$(arg arg_enable_point_cloud_publish)"  />
            <param name="param_encoding_type" value="$(arg arg_encoding_type)" />
        </node>

        <!-- Do the transforms of optical_frame to camera_base frame -->
        <node pkg="tf2_ros" type="static_transform_publisher" name="$(eval arg('cam1_base_frame_optical') + '_tf')"
                args="0 0 0 $(arg cam_optical_to_base_yaw) $(arg cam_optical_to_base_pitch) $(arg cam_optical_to_base_roll) $(arg cam1_base_frame) $(arg cam1_child_frame)" />

        <!-- Transform for Camera Base frame and MAP -->
        <node pkg="tf2_ros" type="static_transform_publisher" name="$(eval arg('cam1_base_frame') + '_tf')"
          args="$(arg cam1_pos_x) $(arg cam1_pos_y) $(arg cam1_pos_z) $(arg cam1_yaw) $(arg cam1_pitch) $(arg cam1_roll) $(arg cam1_parent_frame) $(arg cam1_base_frame)" />

        <!--Dynamic Reconfigure-->
        <node name="adi_3dtof_adtf31xx_dynparam_node" pkg="dynamic_reconfigure" type="dynparam" args="set_from_parameters adi_3dtof_adtf31xx_node">
            <param name = "ab_threshold" type="int" value="$(arg arg_ab_threshold)" />
            <param name = "confidence_threshold" type="int" value="$(arg arg_confidence_threshold)" />
        </node>

    </group>

</launch>

